### ISSUE TITLE
UnicodeDecodeError if the commiter has non-ascii characters

### ISSUE THREAD

**Issue by sdouche** [git-multimail/git-multimail#114]
Version 1.2dev-d64e85e93bd68f2cfd4d5896d125e7d450bdd7a1 (testing under Gerrit).

A UnicodeDecodeError is raised with my commits. If I change my name to Sebastien (s/é/e), it works.

```
Traceback (most recent call last):
  File "/srv/gerrit/hooks/git_multimail.py", line 3506, in <module>
    main(sys.argv[1:])
  File "/srv/gerrit/hooks/git_multimail.py", line 3498, in main
    run_as_update_hook(environment, mailer, refname, oldrev, newrev, options.force_send)
  File "/srv/gerrit/hooks/git_multimail.py", line 3208, in run_as_update_hook
    push.send_emails(mailer, body_filter=environment.filter_body)
  File "/srv/gerrit/hooks/git_multimail.py", line 3130, in send_emails
    rev.recipients,
  File "/srv/gerrit/hooks/git_multimail.py", line 1838, in send
    self.f.writelines(lines)
  File "/srv/gerrit/hooks/git_multimail.py", line 1463, in generate_combined_email
    for line in self.generate_email(push, body_filter, values):
  File "/srv/gerrit/hooks/git_multimail.py", line 795, in generate_email
    for line in self.generate_email_header(**extra_header_values):
  File "/srv/gerrit/hooks/git_multimail.py", line 1072, in generate_email_header
    self.header_template, **extra_values
  File "/srv/gerrit/hooks/git_multimail.py", line 746, in expand_header_lines
    value = addr_header_encode(value, name)
  File "/srv/gerrit/hooks/git_multimail.py", line 450, in addr_header_encode
    header_name="sebastien.douche@scmfrance.fr"
  File "/usr/lib/python2.7/email/header.py", line 183, in __init__
    self.append(s, charset, errors)
  File "/usr/lib/python2.7/email/header.py", line 267, in append
    ustr = unicode(s, incodec, errors)
```

**Comment by edward-dauvergne** [comment]
Could you check if the changes following branch fixes all the problems:  https://github.com/edward-dauvergne/git-multimail/tree/utf-8-email-support2.  I will perform my own testing later, and then make a merge request.

Cheers,

Edward

**Comment by sdouche** [comment]
@edward-dauvergne no more UnicodeDecodeError but headers are malformed:

```
To: =?utf-8?q?sebastien=2Edouche+toto=40scmfrance=2Efr?=
Subject: branch master updated: test7
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: 8bit
Message-ID: <20150818121330.24172.69732@vdgerrit01.leboncoin.lan>
From: =?utf-8?q?review=40bon-coin=2Enet?=
Reply-To: =?utf-8?b?PT91dGYtOD9xP1M9QzM9QTliYXN0aWVuX0RvdWNoZT89IDxzw6li?=
 =?utf-8?q?astien=2Edouche=40scmfrance=2Efr=3E?=
X-Git-Host: vdgerrit01.leboncoin.lan
X-Git-Repo: test-gerrit
X-Git-Refname: refs/heads/master
X-Git-Reftype: branch
X-Git-Oldrev: 925a4aedacc6ffeec31c56498a458d0830ff6661
X-Git-Newrev: d1f85f627546514817183ad5411e8ac9b3676f01
X-Git-Rev: d1f85f627546514817183ad5411e8ac9b3676f01
X-Git-NotificationType: ref_changed_plus_diff
X-Git-Multimail-Version: 1.2.dev
Auto-Submitted: auto-generated

```

**Comment by edward-dauvergne** [comment]
Hi Sébastien,

Could you check if my fixes at https://github.com/git-multimail/git-multimail/pull/80#issuecomment-133751726 improve things?

Cheers,

Edward

**Comment by sdouche** [comment]
Hi @edward-dauvergne,
I check that tomorrow.

**Comment by sdouche** [comment]
@edward-dauvergne where is right PR?

**Comment by moy** [comment]
There's no PR yet, but there's a branch here: https://github.com/edward-dauvergne/git-multimail/commits/utf-8-email-support3

The history needs a few cleanups, but the overall thing sounds OK to me. Does it work for you?

**Comment by moy** [comment]
@sdouche: did you get to test Edward's code?
@edward-dauvergne: don't forget to clean-up and submit a proper PR.

Looking forward to getting this in master :-).

**Comment by edward-dauvergne** [comment]
I've just been a little busy lately.  I wanted to try to understand the system testing infrastructure a little (the non unit tests), and work out what is happening there.  And to try to find out - still unsuccessfully - how to properly encode unicode within the email address.  The double encoding bug is also a headache I'm looking at.

**Comment by moy** [comment]
Don't hesitate to submit a PR with incomplete tests, we can still add some later. One reason I'm looking forward to merging is that the "Python 3" PR will conflict with yours, I'd rather merge yours first and resolve conflicts myself. With this, I'll get to "no pending PR", which didn't happen since a long time ago ;-).

We do need more tests for encoding stuff anyway (e.g. #124).

And don't hesitate to ask questions. The test infrastructure is still a bit hard to get into, but I try to reply to questions with commits to simplify it.

**Comment by edward-dauvergne** [comment]
Ok, I'll clean up what is already in the branch ASAP (it might take a day or two).

**Comment by sdouche** [comment]
@moy tried:
```
File "/srv/gerrit/hooks/git_multimail.py", line 1790, in send
     raise e
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position
```

And you forgot the option 'gerrit'

**Comment by edward-dauvergne** [comment]
Strange, that error should now be impossible as everything is pre-converted to unicode.  Maybe it is worth testing the cleaned up branch at https://github.com/git-multimail/git-multimail/pull/127.  As for 'gerrit', I'm not sure what that is.

**Comment by sdouche** [comment]
@moy forgot to add 'gerrit' in the option parsing (--env).

**Comment by moy** [comment]
About 'gerrit': fixed in b977b3e. But you don't really need it, the environment autodetection just works for Gerrit.

@sdouche : where exactly is the accent in your email? Is it in the address part (like Sébastien@example.com) or in the name part (like Sébastien <sebastien@example.com>)?

**Comment by edward-dauvergne** [comment]
I just ran a live test on a repository using the code now on master.  For the test, my ~/.gitconfig file looks like:
```
[edward@localhost ~]$ cat ~/.gitconfig
[color]
        status = auto
        branch = auto
        diff = auto
[user]
        name = Edward d'Auvergné
        email = edward <att> nmr-relax.com
[push]
        default = simple
[edward@localhost ~]$
```
The email address is deliberately abstracted.  After a local commit, then push, the email arrived with the correct header line:
```
From: =?utf-8?q?Edward_d=27Auvergn=C3=A9?= <edward@nmr-relax.com>
```
In my mail client this is translated and showing correctly as `Edward d'Auvergné <edward <att> nmr-relax.com>`.  Accents in the email address itself - something which is still very, very rare - will not be handled correctly due to a double encoding bug, and due to a different way that this should be encoded.

**Comment by moy** [comment]
>  Accents in the email address itself - something which is still very, very rare

Agreed, but I'd rather hear it explicitly.

My guess is that this is gerrit-related. The gerrit environment gets some information from the command-line, and we may get the encoding differently from the command-line and when reading it from git (either `git config` or `git log`). Or perhaps Gerrit itself encodes the cli options badly?

**Comment by moy** [comment]
I still can't reproduce with python 2:

    ./git-multimail/git_multimail.py --submitter='Sébastien (seb@example.com)' \
        --stdout --oldrev=HEAD^ --newrev=HEAD --refname=refs/heads/master \
        --project='My Project' -c multimailhook.mailingList=foo@example \
        -c multimailhook.from=pusher

I get no exception and properly encoded To: and From: headers. @sdouche : you'll have to provide more details. As much as possible, a script or cut-and-paste ready command to reproduce the problem.

I did manage to get a failure with python 3:

    LC_ALL=C python3 ./git-multimail/git_multimail.py --submitter='Sébastien (seb@example.com)' --stdout --oldrev=HEAD^ --newrev=HEAD --refname=refs/heads/master --project='My Project' -c multimailhook.mailingList=foo@example.com -c multimailhook.from=pusher

But no time to investigate for now.

**Comment by edward-dauvergne** [comment]
I can see the same Python3 failure:

```
Traceback (most recent call last):
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 3683, in <module>
    main(sys.argv[1:])
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 3675, in main
    run_as_update_hook(environment, mailer, refname, oldrev, newrev, options.force_send)
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 3364, in run_as_update_hook
    push.send_emails(mailer, body_filter=environment.filter_body)
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 3294, in send_emails
    change.recipients,
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 1953, in send
    self.f.writelines(lines)
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 869, in generate_email
    for line in self.generate_email_header(**extra_header_values):
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 1185, in generate_email_header
    self.header_template, **extra_values
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 802, in expand_header_lines
    value = addr_header_encode(value, name)
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 495, in addr_header_encode
    for name, emailaddr in getaddresses([text])
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 495, in <genexpr>
    for name, emailaddr in getaddresses([text])
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 479, in header_encode
    return Header(text, header_name=header_name, charset=Charset(charset)).encode()
  File "/usr/lib64/python3.4/email/header.py", line 217, in __init__
    self.append(s, charset, errors)
  File "/usr/lib64/python3.4/email/header.py", line 304, in append
    s.encode(output_charset, errors)
UnicodeEncodeError: 'utf-8' codec can't encode character '\udcc3' in position 1: surrogates not allowed
```
A bit of testing shows that this is a new Python3 problem, as the unicode text 'S\udcc3\udca9bastien <seb@example.com>' is being sent into the `addr_header_encode()` function.  So this is not related to my recent changes.  This requires more investigation.

**Comment by edward-dauvergne** [comment]
To isolate the Python3 problem with the command line --submitter option, try the following patch:

```diff
diff --git a/git-multimail/git_multimail.py b/git-multimail/git_multimail.py
index ead69b0..2080211 100755
--- a/git-multimail/git_multimail.py
+++ b/git-multimail/git_multimail.py
@@ -3623,6 +3623,8 @@ def main(args):
 
     (options, args) = parser.parse_args(args)
     (options, args, hook_info) = check_hook_specific_args(options, args)
+    print(repr(options.submitter))
+    print(options.submitter)
 
     if options.version:
         sys.stdout.write('git-multimail version ' + get_version() + '\n')
```
This catches it at the start - with the Python command line option parsing of `optparse.OptionParser`:
```
$ LC_ALL=C python3 /data/software/git-multimail/git-multimail/git_multimail.py --submitter='Sébastien (seb@example.com)' --stdout --oldrev=HEAD^ --newrev=HEAD --refname=refs/heads/master --project='My Project' -c multimailhook.mailingList=foo@example.com -c multimailhook.from=pusher
'S\udcc3\udca9bastien <seb@example.com>'
Traceback (most recent call last):
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 3681, in <module>
    main(sys.argv[1:])
  File "/data/software/git-multimail/git-multimail/git_multimail.py", line 3627, in main
    print(options.submitter)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 1-2: ordinal not in range(128)
```
Or simply:
```
$ python3
Python 3.4.3 (default, Jul  1 2015, 18:38:11) 
[GCC 4.9.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 'S\udcc3\udca9bastien <seb@example.com>'.encode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'utf-8' codec can't encode character '\udcc3' in position 1: surrogates not allowed
>>> 
```

**Comment by edward-dauvergne** [comment]
This is quite strange, but together with the patch
```diff
diff --git a/git-multimail/git_multimail.py b/git-multimail/git_multimail.py
index ead69b0..d65b5a4 100755
--- a/git-multimail/git_multimail.py
+++ b/git-multimail/git_multimail.py
@@ -3507,6 +3507,9 @@ def compute_gerrit_options(options, args, required_gerrit_options):
         # The submitter argument is almost an RFC 2822 email address; change it
         # from 'User Name (email@domain)' to 'User Name <email@domain>' so it is
         options.submitter = options.submitter.replace('(', '<').replace(')', '>')
+        # Convert to unicode for Python3.
+        if PYTHON3:
+            options.submitter = options.submitter.encode('utf-8', 'surrogateescape').decode('utf-8', 'replace')
     else:
         update_method = 'submitted'
         # Gerrit knew who submitted this patchset, but threw that information
```
adding the `PYTHONIOENCODING=utf-8` option on the command line:
```sh
$ LC_ALL=C PYTHONIOENCODING=utf-8 python3 /data/software/git-multimail/git-multimail/git_multimail.py --submitter='Sébastien (seb@example.com)' --stdout --oldrev=HEAD^ --newrev=HEAD --refname=refs/heads/master --project='My Project' -c multimailhook.mailingList=foo@example.com -c multimailhook.from=pusher
```
allows for unicode characters (or iso-8859-1) in the git-multimail --submitter option!  This must be because `sys.stdout` is converting to ASCII.

**Comment by sdouche** [comment]
Gentlemen, it seems ok with the 83466bfd7f6738d091b1d2926a718e077bb4de67 version.

**Comment by moy** [comment]
I guess the fix is 0d9f54339349d48fbe0df645f2cb5da637d62591. This is weird, since we only got the error with Python 3, but glad to see that it did the trick for you.

At least, the port to Python 3 was useful for something!
